pip/pip-344.md

```diff
@@ -0,0 +1,78 @@
+- For case 5 of `pulsarClient.getPartitionsForTopic`'s behavior, it returns a `0` partitioned metadata, but the topic does not exist. For the correct behavior of this case, we had discussed [here](https://github.com/apache/pulsar/issues/8813) before.
+
+# Goals
+
+Correct the behaviors of case 4 and case 5.
+
+- Do not create the partitioned metadata when calling `pulsarClient.getPartitionsForTopic`. The partitioned metadata will only be created when consumers/producers are trying to register.
+- Instead of returning a `0` partitioned metadata, respond to a not found error when calling `pulsarClient.getPartitionsForTopic` if the topic does not exist.
+
+# Detailed Design
+
+## Public-facing Changes
+
+When you call the public API `pulsarClient.getPartitionsForTopic`, pulsar will not create the partitioned metadata anymore.
+- [flink-connector-pulsar](https://github.com/apache/flink-connector-pulsar/blob/main/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/topic/ProducerRegister.java#L221-L227) is using this API to create partitioned topic metadata.
+
+### Public API
+**LookupService.java**
+```
+// This API existed before. Not change it, thus ensuring compatibility.
+CompletableFuture<PartitionedTopicMetadata> getPartitionedTopicMetadata(TopicName topicName);
+
++ // A new API that contains an additional param "createIfAutoCreationEnabled."
++ CompletableFuture<PartitionedTopicMetadata> getPartitionedTopicMetadata(TopicName topicName, boolean createIfAutoCreationEnabled);
+```
+
+| case | do create partitioned metadata | non-partitioned topic | partitioned topic |  current behavior |
+| --- | --- | --- | --- | --- |
+| 1 | `true/false` | `exists: true` | | REST/Client API: `partitions: 0` |
+| 2 | `true/false` | | `exists: true` <br> `partitions: 3` | REST/Client API: `partitions: 3` |
+| 3 | `true` | | | REST/Client API: <br> &nbsp;&nbsp;- `create new: true` <br> &nbsp;&nbsp;- `partitions: 3` |
+| 4 | `false` | | | REST/Client API: <br> &nbsp;&nbsp;- Not found error |
+
+### Binary protocol
+
+**CommandPartitionedTopicMetadata**
+```
+message CommandPartitionedTopicMetadata {
+  + optional bool metadata_auto_creation_enabled = 6 [default = true];
+}
+```
+
+# Backward & Forward Compatibility
+
+- Old version client and New version Broker
+  - Pulsar will create the partitioned topic metadata when calling `pulsarClient.getPartitionsForTopic`.
+  - The client does not pass the `create_if_auto_creation_enabled` parameter of `CommandPartitionedTopicMetadata` to the Broker, which defaults to `true` and automatically creates the partitioned topic metadata.
+
+- New version client and Old version Broker
+  - Pulsar will create the partitioned topic metadata when calling `pulsarClient.getPartitionsForTopic`.
+  - The client passes the `create_if_auto_creation_enabled` parameter of `CommandPartitionedTopicMetadata` to the Broker; the Broker will discard it and create the partitioned topic metadata automatically.
```

Comment:
When a client user calls the API using `getPartitionedTopicMetadata(myTopic, false)` and the broker is an old version, the topic can still be created. This would break the API semantics and need to make more sense.

Maybe we need to throw an exception here if the broker doesn't support it.

Reply:
> When a client user calls the API using getPartitionedTopicMetadata(myTopic, false) and the broker is an old version, the topic can still be created. This would break the API semantics and need to make more sense.
> Maybe we need to throw an exception here if the broker doesn't support it.

Good pointer. Added a feature flag `supports_binary_api_get_partitioned_meta_with_param_created_false` to prevent the new client from calling a new API with an old broker.

Reply:
@RobertIndie 

The clients will always call the old API if one of the `client` and `broker` does not support the new API.